dask/array/core.py

```diff
@@ -815,20 +815,24 @@ class Array(DaskMethodsMixin):
     --------
     dask.array.from_array
     """
-    __slots__ = 'dask', '_name', '_cached_keys', '_chunks', 'dtype'
+    __slots__ = 'dask', '_name', '_cached_keys', '_chunks', '_meta'
 
-    def __new__(cls, dask, name, chunks, dtype, shape=None):
+    def __new__(cls, dask, name, chunks, dtype=None, meta=None, shape=None):
         self = super(Array, cls).__new__(cls)
         assert isinstance(dask, Mapping)
         if not isinstance(dask, HighLevelGraph):
             dask = HighLevelGraph.from_collections(name, dask, dependencies=())
         self.dask = dask
         self.name = name
-        if dtype is None:
-            raise ValueError("You must specify the dtype of the array")
-        self.dtype = np.dtype(dtype)
+        if dtype is None and meta is None:
+            raise ValueError("You must specify the meta or dtype of the array")
+        if meta is None and dtype is not None:
+            nd = len(shape) if shape is not None else 1
```

Comment:
Hrm, we could use `len(self._chunks)` reliably after that is assigned.  It's unfortunate that that seems to rely on dtype.

Reply:
Maybe something like the following

```python
self._chunks = normalize_chunks(chunks, shape, dtype=dtype or meta.dtype)
meta = np.empty(shape=(0,) * self.ndim, dtype=dtype)
```